File: src/main/java/calculadora/command/Invoker.java
package calculadora.command;

/**
 *
 * @author vedz11
 */
import calculadora.memento.*;

public class Invoker {
    private final Calculator calculator;
    private final History history;

    public Invoker(Calculator calculator, History history) {
        this.calculator = calculator;
        this.history = history;
    }

    public void executeCommand(Command command) {
        // Guardar estado previo antes de ejecutar
        history.save(calculator.save());
        command.execute();
    }

    public void undo() {
        if (history.canUndo()) {
            CalculatorMemento m = history.undo(calculator);
            calculator.restore(m);
        } else {
            System.out.println("No hay acciones para deshacer.");
        }
}

    public void redo() {
        if (history.canRedo()) {
            CalculatorMemento m = history.redo(calculator);
            calculator.restore(m);
        } else {
            System.out.println("No hay acciones para rehacer.");
        }
    }

}

File: src/command/Invoker.ts
import { Calculator } from "../memento/Calculator";
import { History } from "../memento/History";
import type { Command } from "./Command";

export class Invoker {
    private calculator: Calculator;
    private history: History;

    constructor(calculator: Calculator, history: History) {
      this.calculator = calculator;
      this.history = history;
    }

    executeCommand(command: Command): void {
      this.history.save(this.calculator.save());
      command.execute();
    }

    undo(): void {
      if (this.history.canUndo()) {
        const m = this.history.undo(this.calculator);
        if (m) this.calculator.restore(m);
      }
    }

    redo(): void {
      if (this.history.canRedo()) {
        const m = this.history.redo(this.calculator);
        if (m) this.calculator.restore(m);
      }
    }
}